packages/assets-controllers/src/multicall.ts

```diff
@@ -416,3 +466,402 @@ export const multicallOrFallback = async (
 
   return await fallback(calls, maxCallsParallel);
 };
+
+/**
+ * Execute multiple contract calls using Multicall3's aggregate3 function.
+ * This allows for more efficient batch calls with individual failure handling.
+ *
+ * @param calls - Array of calls to execute via aggregate3
+ * @param chainId - The hexadecimal chain id
+ * @param provider - An ethers rpc provider
+ * @returns Promise resolving to array of results from aggregate3
+ */
+export const aggregate3 = async (
+  calls: Aggregate3Call[],
+  chainId: Hex,
+  provider: Web3Provider,
+): Promise<Aggregate3Result[]> => {
+  if (calls.length === 0) {
+    return [];
+  }
+
+  const multicall3Address = MULTICALL_CONTRACT_BY_CHAINID[chainId];
+  const multicall3Contract = new Contract(
+    multicall3Address,
+    multicall3Abi,
+    provider,
+  );
+
+  return await multicall3Contract.callStatic.aggregate3(calls);
+};
+
+/**
+ * Creates ERC20 balance calls for multiple token and user address combinations
+ *
+ * @param tokenAddresses - Array of ERC20 token contract addresses
+ * @param userAddresses - Array of user addresses to check balances for
+ * @param provider - An ethers rpc provider
+ * @returns Object containing calls array and callMapping array for tracking
+ */
+const createERC20BalanceCalls = (
+  tokenAddresses: string[],
+  userAddresses: string[],
+  provider: Web3Provider,
+): {
+  calls: Aggregate3Call[];
+  callMapping: { tokenAddress: string; userAddress: string }[];
+} => {
+  const calls: Aggregate3Call[] = [];
+  const callMapping: { tokenAddress: string; userAddress: string }[] = [];
+
+  // Create a temporary contract instance to encode the balanceOf function calls
+  const tempContract = new Contract(
+    ZERO_ADDRESS,
+    ERC20_BALANCE_OF_ABI,
+    provider,
+  );
+
+  tokenAddresses.forEach((tokenAddress) => {
+    userAddresses.forEach((userAddress) => {
+      calls.push({
```

Comment:
Nice!!

Do we know if there is a limit to the number of call data we can use?
1. Maybe is there an RPC max size request?
2. Does the multicall contract have a max request size?